### PR TITLE
Convert transaction committer to applicative style

### PR DESCRIPTION
**Pull request by mziolekda**
Convert transaction committer to applicative style.

In the original implementation, the computations inside validation were represented as a state monad. The use of the bind operator (flatMap) was resulting in a recursive call and a stack-overflow for large transactions:

```
final case class Commit[A](run: CommitContext => Either[CommitDone, (A, CommitContext)]) {
    def flatMap[A1](f: A => Commit[A1]): Commit[A1] =
      Commit { state =>
        run(state) match {
          case Left(done) => Left(done)
          case Right((x, state2)) =>
            f(x).run(state2)    <<<=== recursion here
        }
      }
```

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 1]
```diff
@@ -139,134 +132,117 @@ private[kvutils] class ProcessTransactionSubmission(
...
+              ctx.getRecordTime,
+              buildRejectionLogEntry(digest, RejectionReason.Disputed(err.msg))),
+          succ => buildFinalResult(ctx, digest, succ))
+
+  private def validateContractKeys: Step = (ctx, digest) => {
+    val damlState = ctx.inputs.collect { case (k, Some(v)) => k -> v } ++ ctx.getOutputs
+    val startingKeys = damlState.collect {
+      case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty => k
+    }.toSet
+    validateContractKeyUniqueness(ctx.getRecordTime, digest, startingKeys) match {
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:233]: Might be worth implementing `flatMap` on `StepResult`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 2]
```diff
@@ -1,35 +1,46 @@
...
 import com.daml.lf.value.Value.AbsoluteContractId
 import com.daml.metrics.Metrics
 import com.google.protobuf.{Timestamp => ProtoTimestamp}
-import org.slf4j.{Logger, LoggerFactory}
 
 import scala.collection.JavaConverters._
 
+private[kvutils] object TransactionCommitter {
+
+  case class Digest(submission: DamlTransactionEntry) {
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:32]:
```suggestion
  case class DamlTransactionEntrySummary(submission: DamlTransactionEntry) {
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 3]
```diff
@@ -1,35 +1,46 @@
...
+    val ledgerEffectiveTime: Timestamp = parseTimestamp(submission.getLedgerEffectiveTime)
+    val submitterInfo: DamlSubmitterInfo = submission.getSubmitterInfo
+    val commandId: String = submitterInfo.getCommandId
+    val submitter: Party = Party.assertFromString(submitterInfo.getSubmitter)
+    lazy val abs: AbsTransaction = Conversions.decodeTransaction(submission.getTransaction)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:37]:
```suggestion
    lazy val absoluteTransaction: AbsTransaction = Conversions.decodeTransaction(submission.getTransaction)
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 4]
```diff
@@ -41,81 +52,63 @@ import scala.collection.JavaConverters._
...
-        "Authorize and build result" ->
-          authorizeAndBlind(recordTime, transactionEntry).flatMap(
-            buildFinalResult(entryId, recordTime, transactionEntry))
-      )
-    }
+) extends Committer[DamlTransactionEntry, TransactionCommitter.Digest] {
+  override protected val committerName = "transaction"
+
+  override protected def init(
+      ctx: CommitContext,
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:64]:
```suggestion
      commitContext: CommitContext,
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 5]
```diff
@@ -139,134 +132,117 @@ private[kvutils] class ProcessTransactionSubmission(
...
-      inputState.collect {
-        case (key, Some(value))
-            if value.hasContractState
-              && value.getContractState.hasContractKey
-              && contractIsActiveAndVisibleToSubmitter(transactionEntry, value.getContractState) =>
-          value.getContractState.getContractKey -> Conversions.stateKeyToContractId(key)
-      }
+  private def validateModelConformance: Step =
+    (ctx, digest) =>
+      metrics.daml.kvutils.committer.transaction.interpretTimer.time(() => {
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:178]: Can this be shortened?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 3 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:37]: Why does it help for it to be `lazy`?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 5 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:178]: I find it quite handy because one can see the full metric patch.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:178]: True, I was thinking about aliasing with a shorter but still meaningful name such as `commitInterpreterTime`, the long name would only be a click away. Disavantage: not one click away on Github so, on second thought, I withdraw it.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 3 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:37]: That was there in the original implementation. I am guessing it is to avoid decoding upfront, when the transaction can fail on a more rudimentary validation before

**Comment by gerolf-da** [comment]
Thanks, this brings transaction processing on the same level as the other committers. What's the advantage of the "applicative style" committers over the previous approach?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 6]
```diff
@@ -41,47 +40,34 @@ import scala.collection.JavaConverters._
...
+    TransactionCommitter.DamlTransactionEntrySummary(transactionEntry)
+
+  override protected val steps: Iterable[(StepInfo, Step)] = Iterable(
+    "authorize_submitter" -> authorizeSubmitter,
+    "check_informee_parties_allocation" -> checkInformeePartiesAllocation,
+    "deduplicate" -> deduplicateCommand,
+    "validate_ledger_time" -> validateLedgerTime,
+    "validate_contract_keys" -> validateContractKeys,
+    "validate_model_conformance" -> validateModelConformance,
+    "authorize_and_build_result" -> authorizeAndBlind
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:64]:
```suggestion
    "authorize_and_blind" -> authorizeAndBlind
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala [hunk 7]
```diff
@@ -93,4 +94,27 @@ private[committer] trait Committer[Submission, PartialResult] {
...
 object Committer {
   type StepInfo = String
+
+  def getCurrentConfiguration(
+      defaultConfig: Configuration,
+      inputState: Map[DamlStateKey, Option[DamlStateValue]],
+      logger: Logger): (Option[DamlConfigurationEntry], Configuration) =
+    inputState
+      .getOrElse(
+        Conversions.configurationStateKey,
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala:104]: `Conversions.configurationStateKey` should probably be moved into this object as a (package private) constant -- it does not seem to be related to the rest of the methods in `Conversions`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 8]
```diff
@@ -176,12 +176,14 @@ class KeyValueCommitting private[daml] (
         )
 
       case DamlSubmission.PayloadCase.TRANSACTION_ENTRY =>
-        new ProcessTransactionSubmission(defaultConfig, engine, metrics, inStaticTimeMode)
+        new TransactionCommitter(defaultConfig, engine, metrics, inStaticTimeMode)
           .run(
             entryId,
+            //TODO replace this call with an explicit maxRecordTime from the request once available
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala:182]: Can you not add this TODO here, please? We already have it in many places and this should be rather captured in our backlog.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 9]
```diff
@@ -139,134 +122,127 @@ private[kvutils] class ProcessTransactionSubmission(
...
+          StepContinue(transactionEntry)
         else
           reject(
-            recordTime,
+            commitContext.getRecordTime,
             buildRejectionLogEntry(
               transactionEntry,
               RejectionReason.SubmitterCannotActViaParticipant(
-                s"Party '${transactionEntry.submitter}' not hosted by participant $participantId"))
+                s"Party '${transactionEntry.submitter}' not hosted by participant $commitContext.getParticipantId")
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:136]:
```suggestion
                s"Party '${transactionEntry.submitter}' not hosted by participant ${commitContext.getParticipantId}")
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 1 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala]: I could also implement semigroup :-) I think it is not worth it for just one occurence

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala (hunk 7 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala:104]: It seems as though it is related to other methods in `Conversions` such as `partyStateKey` and `packageStateKey` each of which returns `DamlStateKey` wrapping. This one just happens to be parameterless.
It is also used in many places throughout the codebase. Moving it and making it private would result in code replication.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala (hunk 8 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala:182]: Done

**Comment by fabiotudone-da** [comment]
The ledger API test tool tests are all-green on a DAML-on-InMemory PKV with this SDK, both with and without batching: for me it's :+1:

**Comment by fabiotudone-da** [comment]
> Thanks, this brings transaction processing on the same level as the other committers. What's the advantage of the "applicative style" committers over the previous approach?

@gerolf-da In short: #5910
